[component|collect|report] Add metadata collection

TurboTurtle opened the pull request:
Adds facilities to capture metadata information about a particular sos
run, and save it to any resulting archive as `manifest.json`. This
manifest will include component-specific information that should assist
in analysis of sos runs.

For `report`, this is an extension of existing reporting functionality,
that records information on a per-plugin basis regarding command output
and file collection. However, more information is recorded in this
collection than the JSON or HTML reporting structures that review a
finished archive.

Manifest data is added at the time of collection; for commands this includes
a breakout of command, parameters, full execution syntax, filepath relative
to archive root, and the actual return code of the command, even if non-zero.
For files, the copied files are grouped by the specification set in
`add_copy_spec()` calls, I.E. for any file pattern regex that collects
at least one file, all files collected by that regex are listed.

Additionally, this adds a `tags` parameter to `add_copy_spec()`,
`add_cmd_output()`, and `collect_cmd_output()` that if used will add
tagging information to the manifest entries for those collections. This
is intended to assist in future sos work (E.G. `sos examine`) and third
party utilities in identifying relevant/connected data that is present
in the archive. Note that all collections specified by a single call to
any of these collection methods will inherit the same tags.

For `collect` we now record the cluster type, node list, and a separate
entry for each node in the node list that includes the sos version, node
sos policy, and the archive within the collect tarball that was
collected for that host.

Closes: #1990
Resolves: #2050

Signed-off-by: Jake Hunsaker <jhunsake@redhat.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [x] If this commit closes an existing issue, is the line `Closes: #ISSUENUMBER` included in an independent line?
- [x] If this commit resolves an existing pull request, is the line `Resolves: #PRNUMBER` included in an independent line?

pmoravec commented on sos/report/plugins/__init__.py:
```diff
@@ -479,6 +495,11 @@ def timeout(self):
 
+    def set_timeout_hit(self):
+        self._timeout_hit = True
+        self.manifest.add_field('end_time', time())
```
Doesnt it matter we store in `end_time` a different format (`time.time()`) than for non-timeouted plugins (`datetime.datetime.now()`)?

Is it worth adding also `run_time`?

TurboTurtle replied:
Oh, good catch! Yup, this should all use `datetime`. Adding a `run_time` field sounds good too

pmoravec commented on sos/report/__init__.py:
```diff
@@ -903,15 +916,21 @@ def _collect_plugin(self, plugin):
-                timeout = self.loaded_plugins[plugin[0]-1][1].timeout or None
+                timeout = _plug.timeout or None
+                start = datetime.now()
+                _plug.manifest.add_field('start_time', start)
                 t.result(timeout=timeout)
+                end = datetime.now()
+                _plug.manifest.add_field('end_time', end)
+                _plug.manifest.add_field('run_time', end - start)
```
A note from an English forever-noob: the `[start|end|run]_time` is well human understandable, but logically incoherent with `setup_[start|end|time]`. But I failed to find out labels that are both logically coherent and such straighforwardly understandable :-/ .

TurboTurtle replied:
Yeah I don't know of a better way to phrase it, but at the same time I don't think this will be a real issue seeing as the contributors would be the only ones interested in the `setup_` fields, and once you're familiar with the fact that plugins run `setup()` before collection, it tracks easily (at least in my mind).

pmoravec commented on sos/report/plugins/__init__.py:
```diff
@@ -1368,10 +1409,22 @@ def _collect_cmd_output(self, cmd, suggest_filename=None,
             # as otherwise we will create a blank file in the archive
             if result['status'] in [126, 127]:
-                return result
+                if self.manifest:
+                    self.manifest.commands.append({
+                        'command': cmd.split(' ')[0],
+                        'parameters': cmd.split(' ')[1:],
+                        'exec': cmd,
+                        'filepath': None,
+                        'return_code': result['status'],
+                        'run_time': time() - start,
+                        'tags': tags
+                    })
+                    return result
+
+        end = time() - start
```
code de-duplication: isnt it worth having:

    end = time() - start
    manifest_cmd = { 'command': .. .. .. 'tags': tags }
    if <command_not_found>:
        if self.manifest:
            self.manifest.commands.append(manifest_cmd)
        return result
    ..
    if manifest:
        manifest_cmd['filepath'] = outfn
        self.manifest.commands.append(manifest_cmd)

(maybe with better name of `manifest_cmd` variable)

pmoravec replied:
Also, `end` variable has confusing name (cf. `'run_time': end,`) - could it be renamed to e.g. `runtime`?

TurboTurtle replied:
Ack to both.

TurboTurtle commented:
@pmoravec updated the PR sans the `setup_` field names noted above.

pmoravec commented on sos/archive.py:
```diff
@@ -532,9 +535,24 @@ def cleanup(self):
         self.log_info("finalizing archive '%s' using method '%s'"
                       % (self._archive_root, method))
+        if self.manifest is not None:
+            self.add_final_manifest_data(method)
```
No metadata are written when `--build` option is used. Since this `finalize` method is called only when `--build` isnt set (cf. https://github.com/sosreport/sos/blob/master/sos/report/__init__.py#L1099-L1106).

A possible fix is to call at https://github.com/sosreport/sos/blob/master/sos/report/__init__.py#L1122 either `self.archive.add_final_manifest_data(None)` directly, or `self.archive.finalize(None)` (that would require some updates in `finalize` to skip compression when `method=None`).

TurboTurtle replied:
Yeah, my hope was that putting it in `archive.finalize()` was to "automagic" it for any components that write one, but I overlooked this.

But, at the same time not a big issue to make the call in each component.

pmoravec commented:
Two - rather stylistic - issues that I dont have a strong opinion on: resulting `manifest.json` has two types of duration:

- `"run_time": "0:00:07.225778",` for plugins run time
- `"run_time": 0.07651472091674805,` for commands run time

My comments:
- the high precision of commands run time is redundant (isnt calling such high precision (date)time more expensive, also?)
- we shall have (do we really?) one common format for the run time; but if, so, which one? `H:MM:SS.sssss` has pros as human readable, pure `S.ssss` is better for automatic processing

I dont have strong preferences of either comment and I am fine with the current implementation wrt. this. Just raising as points for a consideration/assessment.

pmoravec commented:
> @pmoravec updated the PR sans the `setup_` field names noted above.

No other comments than those two (`--build`, and runtime format). Good work!

TurboTurtle commented:
> * we shall have (do we really?) one common format for the run time; but if, so, which one? `H:MM:SS.sssss` has pros as human readable, pure `S.ssss` is better for automatic processing
> 

So, the command execution precision was already there - I just lifted it into the manifest data. It's what we currently log, e.g.:

```
[plugin:kernel] collecting output of 'sysctl -a'
[plugin:kernel] collected output of 'sysctl' in 0.034815073013305664 (changes=False)
[plugin:kernel] unpacked command: binary=False, changes=False, chroot=True, cmd='dkms status', env=None, foreground=False, root_symlink=None, runat=None, sizelimit=25, stderr=True, subdir=None, suggest_filename=None, timeout=300
[plugin:kernel] collecting output of 'dkms status'
[plugin:kernel] collected output of 'dkms' in 0.08468127250671387 (changes=False)
[plugin:kernel] collected plugin 'kernel' in 2.217618465423584
```

I put the plugin's `*_time` in datetime format because that is easier to read for humans. I think for now we should move forward with the current implementation, but we can always pivot if this causes issues or confusion in real world runs.

Unless of course anyone has an objection to doing that, in which case just let us know here before merge.

TurboTurtle commented:
I've updated the branch to write the manifest during `final_work()` before we decide what to do with `--build`. Also moved the location to `sos_reports/manifest.json` per IRC.
